pkg/wal/wal.go

```diff
@@ -0,0 +1,345 @@
+	if err != nil {
+		w.l.Error("failed to get tokens",
+			zap.Error(err),
+			zap.String("backDatedToken", ksuidOld.String()),
+			zap.String("token", fromToken))
+	}
+	return tokens, next, err
+}
+
+
+type responseEntries struct {
+	entries *iradix.Txn
+	mutex sync.Mutex
+	err error
+}
+
+func (r *responseEntries) addResponse(entry *Entry) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	_, update := r.entries.Insert([]byte(entry.Token), entry)
+	if update {
+		panic("tried to enter the same entry more than once: "+ entry.Token)
+	}
+}
+
+func (r *responseEntries) listEntries() []Entry {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	iterator := r.entries.Root().Iterator()
+	var entries []Entry
+	for {
+		b, v, ok := iterator.Next()
+		if !ok {
+			break
+		}
+		key := string(b)
+		entry, ok := v.(Entry)
+		if !ok {
+			panic("Got a type other than Entry: "+key)
+		}
+		entries = append(entries, entry)
+	}
+	return entries
+}
+
+func (w *WAL) getConnection() {
+	w.connectionControl <- struct{}{}
+	return
+}
+
+func (w *WAL) releaseConnection() {
+	<- w.connectionControl
+	return
+}
+
+type walChannels struct {
+	tokens  chan []string
+	entry   chan *Entry
+	entries chan []Entry
+	count   chan int
+	oops    chan error
+	done chan struct{}
+}
+
+func newWalChannels() *walChannels {
+	token := make(chan []string)
+	entry := make(chan *Entry)
+	entries := make(chan []Entry)
+	count := make (chan int)
+	done := make(chan struct{})
+	return &walChannels{
+		tokens:  token,
+		entry:   entry,
+		entries: entries,
+		count:   count,
+		done:    done,
+	}
+}
+
+func (w *WAL) ListEntries(ctx context.Context, fromToken string, max int) ([]Entry, string,  error) {
+
+	walChannels := newWalChannels()
+	count := 0
+
+	// Start routine which collects all parallel responses.
+	go w.collectParallelResponses(ctx, walChannels)
+	// Start routine that will issue parallel reads.
+	go w.issueParallelReads(ctx, walChannels)
+
+	// Read tokens and send to routine to entry in parallel while fetching new tokens.
+	for count = 0; count < max; {
+		var tokens []string
+
+		tokens, next, err := w.ListTokens(ctx, fromToken, max-count)
+
+		if err != nil {
+			return nil, "", fmt.Errorf("wal failed to ListEntries: " + err.Error())
+		}
+
+		count += len(tokens)
+
+		if len(tokens) > 0 {
+			// account for case list is not complete but the response does not include any token.
+			walChannels.tokens <- tokens
+		}
+
+		fromToken = next
+
+		// Read max or all
+		if (count == max) || (fromToken == "") {
+			break
+		}
+	}
+
+	close(walChannels.tokens) // No more tokens to report.
+
+	walChannels.count <- count // Send total count for reading responses.
+
+	for {
+		select {
+		case entries := <- walChannels.entries:
+			if len(entries) != count {
+				panic(fmt.Sprintf("received count different than the list of tokens entries: %d count: %d", len(entries), count))
+			}
+			return entries, fromToken, nil
+		case err := <- walChannels.oops:
+			return nil, "", err
```

Comment:
ok so maybe add a comment here that we are willingly leaking routines

Reply:
I am not sure we are leading here, there is only one goroutine that reports into these channels and it will report success or a failure. ```collectParallelResponses``` collects parallel responses and at the end reports one of the errors (After logging them all) or reports a success.

Reply:
```
// Start routine which collects all parallel responses.
	go w.collectParallelResponses(ctx, walChannels)
	// Start routine that will issue parallel reads.
	go w.issueParallelReads(ctx, walChannels)
```
are not waited for when returning, which is what I am calling leaking routines.
This is fine by me provided you make it explicit that this is on purpose and not a bug by omission.

Reply:
Added a note.